book/src/developers/designs/replication_coordinator.md

```diff
@@ -133,160 +106,217 @@ kanidmd replication get-certificate
-configuration, so that the KRC can later issue out which nodes are responsible for those processes.
-
-These are analogous to the AD FSMO roles, but I think we need a different name for them. Single Node
-Origin Task? Single Node Operation Runner? Yes I'm trying to make silly acronyms.
-
 ### KRC Configuration
 
-> Still not fully sure about the KRC config yet. More thinking needed!
+The KRC is implied as a server node by the *lack* of a `krc_client` configuration. This informs
+the node that it must not contact other nodes for it's replication topology, and it prepares the node
+for serving that replication metadata. This is analgous to a single node operation configuration.
 
-The KRC is configured with it's URL and certificates.
+```
+[replication]
+origin = "repl://kanidmd_a:8444"
+bindaddress = "[::]:8444"
+# krc_url
+# krc_ca_dir
+```
+
+All other nodes will have a configuration of:
 
 ```
-[krc_config]
-origin = https://krc.example.com
-tls_chain = /path/to/tls/chain
-tls_key = /path/to/tls/key
+[replication]
+origin = "repl://kanidmd_b:8444"
+bindaddress = "[::]:8444"
+
+# krc_url = https://private.name.of.krc.node
+krc_url = https://kanidmd_a
+# must contain ca that signs kanidmd_a's tls_chain.
+krc_ca_dir = /path/to/ca_dir
 ```
 
-The KRC is also configured with replication groups.
+The domain will automatically add a `Default Site`. The KRC implies it's own membership to "Default Site"
+and it will internally add itself to the`Default Site`
+
+The KRC can then issue Tokens that define which Site a new replica should join. Initially we will
+only allow `Default Site` (and will disallow creation of other sites).
+
+The new replica will load it's krc token from the environment variable `KANIDMD_KRC_TOKEN_PATH`. This
+value will contain a file path where the JWT is stored. This is compatible with systemd credentials and
+docker secrets. By default the value if unset will be defined by a profile default
+(`/etc/kanidm/krc.token` or `/data/krc.token`).
+
+A new replica can then contact the `krc_url` validating the presented TLS chain with the roots
+from `krc_ca_dir` to assert the legitimacy of the KRC. Only once these are asserted, then the
+KRC token can be sent to the instance as a `Bearer` token. The new replica will also provide it's
+mTLS certificate and it's server uuid.
+
+
+Once validated, the KRC will create or update the server's replica entry. The replica entry
+in the database will contain the active mTLS cert of the replica and a reference to the replication
+site that the token referenced.
+
+This will additionally add the "time first seen" to the server entry.
+
+From this, for each server in the replication site associated to the token, the KRC will provide
+a replication config map to the new replica providing all URL's and mTLS certs.
+
+
+Anytime the replica checks in, if the KRC replication map has changed a new one will be provided,
```

Comment:
Why not just return the "current" map every time with an `ETag` header (usually used to identify changes), and the client can decide what to do with it... then if it reboots or loses its copy, it can parse and update its internal representation of the state.

This means it's simpler on the KRC side (not tracking who has what) and the client *always* looks at the replication config and goes "hey is this the same as mine".

Generation IDs make sense just from a change tracking perspective on the human/audit side, but there's no loss in returning a relatively small set of data periodically to KRC clients.

Reply:
If the client has lost it's replication map it'll need to be refreshed and reset anyway in this case because it's then ambiguous if there is other dataloss.

Reply:
Yeah, I'm just thinking it'll be simpler on the server side to make/cache a config and return that every time, less code more betterer

Reply:
Yeah, but network bandwidth also isn't infinite in replication and bigger topologies means we're sending big config maps all the time. The other part is that at some point the client or the server need to choose "when do I need to change and reload my config". I can see it going either way, it's not set in stone really. I just think it's easier for the server to calc that change since it processes all the changes to replication topology, and can work out "okay here is the info you need" rather than the client.

Reply:
'k